native/lib.rs

```diff
@@ -309,8 +325,12 @@ impl PartialEq for HBin {
 
 impl Eq for HBin {}
 
+// All caches are stored in this global map which maps `cache name atom => cache`.
 lazy_static! {
-    static ref CACHES: Mutex<HashMap<HBin, TermCache>> = Mutex::new(HashMap::new());
+    //                                cache name atom
+    //                                |     cache
+    //                                |     |
+    static ref CACHES: RwLock<HashMap<HBin, Arc<RwLock<TermCache>>>> = RwLock::new(HashMap::new());
```

Comment:
Viewing the code via Emacs with rustc **1.62.0** released earlier this week, it complains:
> `*mut u8` cannot be shared between threads safely

plus offers to help:
> within `(HBin, Arc<RwLock<TermCache>>)`, the trait `Sync` is not implemented for `*mut u8`

Reply:
Yep. That's expected

Reply:
I've temporary implemented Sync for the HBin wrapper type to allow the code compile and run. But keep in mind that I'm still not sure if it's sound to do so.

Reply:
Considering the "potential deadlock example" in the doc for RwLock, maybe move the Arc.
How about `RwLock<Arc<HashMap<Hbin, RwLock<TermCache>>>`?

Reply:
> Considering the "potential deadlock example" in the doc for RwLock, maybe move the Arc.

Can you please expand on that?

Reply:
For soundness, we mean avoid deadlock while reducing lock contention:
- Having nested `RwLocks` in the order that you have them in the new code should reduce that contention, but we still must avoid deadlock.
- `Arc` repositioned to be around `HashMap` allows concurrent read of `HBin` ("cache atom name") AND `RwLock` of cache value.
- However, an `Arc` around just the cache value omits concurrency protection for the HashMap itself because `HashMap` doesn't include `Arc` within its private definition.
- The existing inner `Arc` of `CACHES` might be too deep within the structure hierarchy to give desired protection.

Reply:
I'll have to think about your proposal when the fog wears off, but for the time being, keep in mind that deadlock example requires a single thread to attempt read lock twice.

Reply:
Oh, right.  Good catch!

Still, I think the point stands: since RwLock and HashMap don't have an explicit Arc, it would be prudent to put ours within the outermost layer of the `CACHES` structure.  Part of the Rust design philosophy as I understand it: these std library components omit things like Arc in their definitions for allowing us more explicit control over where to put it.

Since this will be called from Erlang in ways where we have less control, it's the 1:1000 (or probably 1:1M) case indicating that this is where `Arc` belongs.

Reply:
Resolving this thread.  We spoke directly since there was a slight disconnect.  

In short: the code within this PR is the best course of action for the criteria and constraints as we understand them to be.